mne/decoding/ssd.py

```diff
@@ -0,0 +1,316 @@
+        self.rank = rank
+        self.cov_method_params = cov_method_params
+
+    def _check_X(self, X):
+        """Check input data."""
+        if not isinstance(X, np.ndarray):
+            raise ValueError('X should be of type ndarray (got %s).'
+                             % type(X))
+        if X.ndim < 2:
+            raise ValueError('X must have at least 2 dimensions.')
+        elif X.ndim > 3:
+            raise ValueError('X must have at most 3 dimensions.')
+
+        n_chan = 0
+        if X.ndim == 2:
+            n_chan = X.shape[0]
+        elif X.ndim == 3:
+            n_chan = X.shape[1]
+        if n_chan != self.info['nchan']:
+            raise ValueError('Info must match the input data.'
+                             'Found %i channels but expected %i.' %
+                             (n_chan, self.info['nchan']))
+
+    def fit(self, X, y=None):
+        """Estimate the SSD decomposition on raw or epoched data.
+
+        Parameters
+        ----------
+        X : array, shape (n_channels, n_times) | shape (n_epochs,
+                n_channels, n_times)
+            The input data from which to estimate the SSD. Either 2D array
+            obtained from continuous data or 3D array obtained from epoched
+            data.
+        y : None | array, shape (n_samples,)
+                    Used for scikit-learn compatibility.
+
+        Returns
+        -------
+        self : instance of SSD
+            Returns the modified instance.
+        """
+        self._check_X(X)
+        if X.ndim == 2:
+            X_aux = X[self.picks_, :].copy()
+
+        elif X.ndim == 3:
+            X_aux = X[:, self.picks_, :].copy()
+        else:
+            raise NotImplementedError()
+
+        X_signal = filter_data(
+            X_aux, self.info['sfreq'], **self.filt_params_signal)
+        X_noise = filter_data(
+            X_aux, self.info['sfreq'], **self.filt_params_noise)
+
+        X_noise -= X_signal
+        if X.ndim == 3:
+            X_signal = np.hstack(X_signal)
+            X_noise = np.hstack(X_noise)
+
+        cov_signal = _regularized_covariance(
+            X_signal, reg=self.estimator, method_params=self.cov_method_params,
+            rank=self.rank, info=self.info)
+        cov_noise = _regularized_covariance(
+            X_noise, reg=self.estimator, method_params=self.cov_method_params,
+            rank=self.rank, info=self.info)
+
+        eigvals_, eigvects_ = eigh(cov_signal, cov_noise)
+        # sort in descending order
+        ix = np.argsort(eigvals_)[::-1]
+        self.eigvals_ = eigvals_[ix]
+        self.filters_ = eigvects_[:, ix]
+        self.patterns_ = np.linalg.pinv(self.filters_)
+
+        return self
+
+    def transform(self, X, y=None):
+        """Estimate epochs sources given the SSD filters.
+
+        Parameters
+        ----------
+        X : array, shape (n_channels, n_times) | shape (n_epochs,
+                n_channels, n_times)
```

Comment:
```suggestion
        X : array, shape ([n_epochs, ]n_channels, n_times)
```

Reply:
thanks! clever way of summarizing both types of array dim ;)